AbstractSQLCompiler.ts

```diff
@@ -359,15 +367,164 @@ const compileSchema = (abstractSqlModel: AbstractSqlModel, engine: Engines, ifNo
 	}
 }
 
+const genSplit = <T>(src:Array<T>, dst: Array<T>, matchFn: MatchFn<T>): Split<T> => {
+	let mod: Array<Pair<T>> = []
+	return _.reduce(src, (acc, value) => {
+		const match = matchFn(acc.ins, value)
+		if (_.isUndefined(match)) {
+			return acc
+		} else {
+			acc.ins = _.without(acc.ins, match)
+			acc.del = _.without(acc.del, value)
+			acc.mod.push( { src: value, dst: match } )
+			return acc
+		} 
+	}
+	, { ins: dst, del: src, mod: mod } )
+}
+//  genDiff expects src and dst arrays formatted as described in
+//  genSplit, and three functions to be called in the case of
+//  insertion, deletion or modification.
+//  All these functions are treated as reducers, their first argument
+//  will be an accumulator of all the changes up to that point and
+//  they are expected to add their changes to it and return it.
+//  The second argument will be an array of [key, value] for
+//  insertions and deletions and [key, pair] for the modifications.
+//  The match function describes how to find the corresponding T inside a T[].
+//  The result of the match function will be either the element that should be matched to 
+//  the argument or undefined if no such element is found.
+const genDiff = <T>(insFn: ReduceFn<T>, delFn: ReduceFn<T>, modFn: ReduceFn<Pair<T>>, matchFn: MatchFn<T>, src: Array<T>, dst: Array<T>) => {
+	let split = genSplit(src, dst, matchFn)
+	return  _.flatten(_.reduce(split.ins, insFn, 
+						_.reduce(split.del, delFn,
+							_.reduce(split.mod, modFn, []))))
+}
+
+const diffFields = (src: AbstractSqlField[], dst: AbstractSqlField[], mappings: ResourceMap<string>, engine: Engines, ifNotExists: boolean) => {
+	let ifNotExistsStr: string
+	let ifExistsStr: string
+
+	if (ifNotExists) {
+		ifNotExistsStr = 'IF NOT EXISTS '
+		ifExistsStr = 'IF EXISTS ' 
+	} else {
+		ifNotExistsStr = ''
+		ifExistsStr = ''
+	}
+	//  defaultTypes =
+	//  	'Short Text': '\'\''
+	//  	'Text': '\'\''
+	//  	'Integer': 0
+
+	let matchFn: MatchFn<AbstractSqlField> = (fieldArray, field) => {
```

Comment:
I'm pretty sure this has an `any` return type, can you check and type it correctly

Reply:
I am not sure I understand. Both supplied matchFn typecheck correctly for me and the type definition specifies the return must be `T | undefined`

Reply:
Ah, you're right the return function will be taken from `MatchFn<AbstractSqlField>`, my mistake